### PR TITLE
prevent URLs in attributes being escaped

### PR DESCRIPTION
**Pull request by alexnguyennz**
## Changes

[Issue](https://github.com/withastro/astro/issues/9695)

Adds check for server rendered URLs attributes with any `&` so they're not escaped. For edge cases where multiple params are needed e.g. for an API that generates OG images.

Before:
`<meta name="og:image" content="https://example.com/api/og?title=hello&#38;description=somedescription">`

After:
`<meta name="og:image" content="https://example.com/api/og?title=hello&description=somedescription">`

I think this change is safe - let me know if there's another way this should be done.

## Testing

`pnpm run test`

And manually with `minimal` demo in dev and build with:

```
<head>
  <meta
    name="og:image"
    content={'https://example.com/api/og?title=hello&description=somedescription'}
  />
</head>
```

## Docs

N/A

### CODE REVIEW

File: .changeset/quick-islands-ring.md
Comment (florian-lefebvre) [inline .changeset/quick-islands-ring.md:5]:
```suggestion
Prevents URLs in attributes from being escaped in static builds
```
I think it concerns any attribute, eg. hrefs

File: packages/astro/src/runtime/server/render/util.ts
Comment (florian-lefebvre) [inline packages/astro/src/runtime/server/render/util.ts:108]: Not familiar with this part of the codebase, but can  you for hrefs attributes as well?

**Comment by bluwy** [comment]
Do we need to escape the `&` in the first place? It should be fine to be used in attribute values 🤔  Seems like it's added since https://github.com/withastro/astro/pull/1789. Perhaps @natemoo-re has some context about it.

File: .changeset/quick-islands-ring.md
Comment (alexnguyennz) [reply .changeset/quick-islands-ring.md]: Yeah I think you're right. I've just set it as a general URL check for any attribute for now.

**Comment by natemoo-re** [comment]
> Do we need to escape the `&` in the first place? It should be fine to be used in attribute values 🤔 Seems like it's added since #1789. Perhaps @natemoo-re has some context about it.

Good question! Escaping `&` is best practice. Seems like [Preact](https://github.com/preactjs/preact-render-to-string/blob/22cb8461a70d86e0e3e4bf36320812a50591814a/src/util.js#L8) and [Svelte](https://github.com/sveltejs/svelte/blob/d57eff76ed24ae2330f11f3d3938761ae4e14b4b/packages/svelte/src/internal/server/index.js#L27) escape it by default.

I'm not sure that we should hard-code the `content` attribute... Maybe we could check for the existence of `&` and then check for a URL? I'm not sure of a simple way we can determine if the & should be escaped or if it's potentially needed for a URL.

**Comment by alexnguyennz** [comment]
> I'm not sure that we should hard-code the `content` attribute... Maybe we could check for the existence of `&` and then check for a URL? I'm not sure of a simple way we can determine if the & should be escaped or if it's potentially needed for a URL.

Not sure either - it's now just a general URL check (with any &) for all attributes if that works.

**Comment by bluwy** [comment]
> Good question! Escaping `&` is best practice. Seems like [Preact](https://github.com/preactjs/preact-render-to-string/blob/22cb8461a70d86e0e3e4bf36320812a50591814a/src/util.js#L8) and [Svelte](https://github.com/sveltejs/svelte/blob/d57eff76ed24ae2330f11f3d3938761ae4e14b4b/packages/svelte/src/internal/server/index.js#L27) escape it by default.

Thanks! It seems like they encode `&` as `&amp;` and `"` as `&quot;`, while we encode as `&#38;` and `&#34;` respectively. Perhaps we need to change how we encode this as well? Technically both should be equivalent, but maybe browsers are able to parse `&amp;` better because the `#` is not tripping it up (?).

**Comment by ematipico** [comment]
I also think that using HTML entities is better.

**Comment by ematipico** [comment]
This PR is sitting for a while without too much activity. cc @alexnguyennz  and @natemoo-re

**Comment by natemoo-re** [comment]
Thanks for the ping @ematipico! I'm okay with this solution if everyone else is.

We can follow-up with switching to `&amp;` and `&quot;` in another PR.

File: .changeset/quick-islands-ring.md
Comment (natemoo-re) [inline .changeset/quick-islands-ring.md:5]:
```suggestion
Prevents fully formed URLs in attributes from being escaped
```

**Review by natemoo-re — APPROVED** [review]

File: packages/astro/src/runtime/server/render/util.ts
Comment (natemoo-re) [inline packages/astro/src/runtime/server/render/util.ts:108]: I noticed that [`URL.canParse`](https://nodejs.org/api/url.html#urlcanparseinput-base) was added in Node v18.17.0, but our minimum required version is still v18.14.1.

We should either implement a fallback for `URL.canParse` (`try/catch`) or wait on merging this until we bump our engines to v.18.17.0 (possibly in the next minor cc @Princesseuh)

**Review by natemoo-re — CHANGES_REQUESTED** [review]
Just blocking until we resolve the `URL.canParse` discrepancy.

Nothing for you to do @alexnguyennz, the maintainers just need to decide when we're updating our Node version.

File: packages/astro/src/runtime/server/render/util.ts
Comment (lilnasy) [reply packages/astro/src/runtime/server/render/util.ts:108]: [`URL.canParse`](https://url.spec.whatwg.org/#dom-url-canparse) uses the same internal method as the [`URL constructor`](https://url.spec.whatwg.org/#constructors), so a try-catch should be sufficient to unblock.

**Review by lilnasy — APPROVED** [review]
Looks great to me! It must have taken effort to find the relevant code path, thanks for the clean fix!

File: packages/astro/src/runtime/server/render/util.ts
Comment (Fryuni) [inline packages/astro/src/runtime/server/render/util.ts:235]: Are relative URLs expected to be unescaped as well? If the OG image in the example were only `/api/og?title=hello&description=somedescription` to be relative to the current browser origin, it wouldn't be escaped as that is not a valid URL.

If we want to validate relative URLs as well this should be:
```suggestion
		new URL(url, 'http://example.com');
```

Any base there would work.

Comment (lilnasy) [reply packages/astro/src/runtime/server/render/util.ts:235]: URL.canParse wouldn't have allowed it but this seems like a great enhancement to me, thanks!

Comment (lilnasy) [reply packages/astro/src/runtime/server/render/util.ts]: On second thought, it may be too permissive.
I don't think there's a case anymore where urlCanParse will fail. If the attribute value has a "&", it always prints unescaped.

Comment (Fryuni) [reply packages/astro/src/runtime/server/render/util.ts]: `URL.canParse` can also receive a base as a second argument, so this applies after upgrading the Node version as well

Comment (lilnasy) [reply packages/astro/src/runtime/server/render/util.ts]: You're right, and we might want to be careful with it as well. I'm not sure how valid it is, but the threat is that we let an attribute like "&\\"" through and break everything.

Comment (Fryuni) [reply packages/astro/src/runtime/server/render/util.ts]: > On second thought, it may be too permissive. I don't think there's a case anymore where urlCanParse will fail. If the attribute value has a "&", it always prints unescaped.

Indeed, it is very hard for a string not to be a valid URL when there is a base. It was hard to find a test case that didn't pass it when we had a similar use case.

Makes this check do almost nothing.

Comment (lilnasy) [reply packages/astro/src/runtime/server/render/util.ts]: You're right. Undoing the change.

**Review by natemoo-re — APPROVED** [review]
Thanks for getting this over the line, @lilnasy!

**Pull request merged**
